File: backend/src/main/java/backend/service/CountryService.java
package backend.service;

import backend.entity.Country;
import backend.repository.CountryRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import java.util.List;
import java.util.Optional;

@Service
public class CountryService {

    @Autowired
    private CountryRepository countryRepository;

    /**
     * Get all countries
     */
    public List<Country> getAllCountries() {
        return countryRepository.findAllOrderByName();
    }

    /**
     * Get countries by continent
     */
    public List<Country> getCountriesByContinent(Country.Continent continent) {
        return countryRepository.findByContinent(continent);
    }

    /**
     * Get all continents
     */
    public List<Country.Continent> getAllContinents() {
        return countryRepository.findAllContinents();
    }

    /**
     * Get countries by visa requirement
     */
    public List<Country> getCountriesByVisaRequired(Boolean visaRequired) {
        return countryRepository.findByVisaRequired(visaRequired);
    }

    /**
     * Get country by ID
     */
    public Optional<Country> getCountryById(Long id) {
        return countryRepository.findById(id);
    }

    /**
     * Get country by code
     */
    public Optional<Country> getCountryByCode(String code) {
        return countryRepository.findByCode(code);
    }

    /**
     * Check if country exists by code
     */
    public boolean existsByCode(String code) {
        return countryRepository.existsByCode(code);
    }

    /**
     * Create new country
     */
    public Country createCountry(Country country) {
        return countryRepository.save(country);
    }

    /**
     * Update country
     */
    public Country updateCountry(Long id, Country countryDetails) {
        Country country = countryRepository.findById(id)
                .orElseThrow(() -> new RuntimeException("Country not found with id: " + id));

        country.setName(countryDetails.getName());
        country.setCode(countryDetails.getCode());
        country.setContinent(countryDetails.getContinent());
        country.setCurrency(countryDetails.getCurrency());
        country.setVisaRequired(countryDetails.getVisaRequired());
        country.setFlagUrl(countryDetails.getFlagUrl());

        return countryRepository.save(country);
    }

    /**
     * Delete country
     */
    public void deleteCountry(Long id) {
        if (!countryRepository.existsById(id)) {
            throw new RuntimeException("Country not found with id: " + id);
        }
        countryRepository.deleteById(id);
    }
}

File: frontend/src/services/countryService.ts
import { apiClient } from './api';

// Country types
export interface Country {
  id: number;
  name: string;
  code: string; // VN, JP, KR, etc.
  continent: 'ASIA' | 'EUROPE' | 'AMERICA' | 'AFRICA' | 'OCEANIA';
  currency?: string;
  visaRequired: boolean;
  flagUrl?: string;
  createdAt: string;
  updatedAt: string;
}

export type Continent = 'ASIA' | 'EUROPE' | 'AMERICA' | 'AFRICA' | 'OCEANIA';

const countryService = {
  // Get all countries
  getAllCountries: async (): Promise<Country[]> => {
    const response = await apiClient.get<Country[]>('/countries');
    return response.data;
  },

  // Get country by ID
  getCountryById: async (id: number): Promise<Country> => {
    const response = await apiClient.get<Country>(`/countries/${id}`);
    return response.data;
  },

  // Get country by code
  getCountryByCode: async (code: string): Promise<Country> => {
    const response = await apiClient.get<Country>(`/countries/code/${code}`);
    return response.data;
  },

  // Get countries by continent
  getCountriesByContinent: async (continent: string): Promise<Country[]> => {
    const response = await apiClient.get<Country[]>(`/countries/continent/${continent}`);
    return response.data;
  },

  // Get all continents
  getAllContinents: async (): Promise<Continent[]> => {
    const response = await apiClient.get<Continent[]>('/countries/continents');
    return response.data;
  },

  // Get countries by visa requirement
  getCountriesByVisaRequirement: async (required: boolean): Promise<Country[]> => {
    const response = await apiClient.get<Country[]>(`/countries/visa-required/${required}`);
    return response.data;
  },

  // Create country (admin only)
  createCountry: async (countryData: Omit<Country, 'id' | 'createdAt' | 'updatedAt'>): Promise<Country> => {
    const response = await apiClient.post<Country>('/countries', countryData);
    return response.data;
  },

  // Update country (admin only)
  updateCountry: async (id: number, countryData: Partial<Omit<Country, 'id' | 'createdAt' | 'updatedAt'>>): Promise<Country> => {
    const response = await apiClient.put<Country>(`/countries/${id}`, countryData);
    return response.data;
  },

  // Delete country (admin only)
  deleteCountry: async (id: number): Promise<void> => {
    await apiClient.delete(`/countries/${id}`);
  },
};

export default countryService;
